project_recalculate/models/project_task.py

```diff
@@ -0,0 +1,253 @@
+        """
+        self.ensure_one()
+        # If no date changes, do nothing
+        if 'date_start' not in vals and 'date_end' not in vals:
+            return vals
+        # If we are changing dates because of task recalculating, do nothing
+        if self.env.context.get('task_recalculate'):
+            return vals
+        date_start = vals.get('date_start', self.date_start)
+        date_end = vals.get('date_end', self.date_end)
+        # If any date is False, can't calculate estimated_days nor from_days
+        if not date_start or not date_end:
+            return vals
+        dec = fields.Datetime.from_string
+        start = dec(date_start)
+        end = dec(date_end)
+        # If end < start, do nothing
+        if end < start:
+            return vals
+        rc = self.pool['resource.calendar']
+        resource, calendar = self._resource_calendar_select()
+        default_interval = self._interval_default_get()
+        calendar_id = calendar.id if calendar else None
+        resource_id = resource.id if resource else None
+        # Calculate estimated_day
+        vals['estimated_days'] = rc.get_working_days_of_date(
+            self.env.cr, self.env.uid, calendar_id, start_dt=start,
+            end_dt=end, compute_leaves=True, resource_id=resource_id,
+            default_interval=default_interval, context=self.env.context)
+        # Calculate from_days depending on project calculation type
+        calculation_type = self.project_id.calculation_type
+        if calculation_type:
+            invert = False
+            increment = calculation_type == 'date_begin'
+            if increment:
+                if not self.project_id.date_start:
+                    # Can't calculate from_days without project date_start
+                    return vals
+                project_date = dec(self.project_id.date_start)
+                start, end = project_date, start
+            else:
+                if not self.project_id.date:
+                    # Can't calculate from_days without project date
+                    return vals
+                project_date = dec(self.project_id.date)
+                start, end = start, project_date
+            if end < start:
+                invert = True
+                start, end = end, start
+            from_days = rc.get_working_days_of_date(
+                self.env.cr, self.env.uid, calendar_id, start_dt=start,
+                end_dt=end, compute_leaves=True, resource_id=resource_id,
+                default_interval=default_interval, context=self.env.context)
+            if invert and from_days:
+                from_days = from_days * (-1)
+            from_days = self._from_days_enc(
+                from_days, project_date, resource, calendar, increment)
+            vals['from_days'] = from_days
+        return vals
+
+    def _estimated_days_prepare(self, vals):
+        # estimated_days must be greater than zero, if not defaults to 1
```

Comment:
This method could be inherited, for example by an addon that calculate it from planned_hours and work_load fields